File: kbmf1mkl1k/kbmf1mkl1k_semisupervised_classification_variational_test.R
# Mehmet Gonen (mehmet.gonen@gmail.com)

kbmf1mkl1k_semisupervised_classification_variational_test <- function(Kx, Kz, state) {
  Nx <- dim(Kx)[2]
  Px <- dim(Kx)[3]
  Nz <- dim(Kz)[2]
  R <- dim(state$Ax$mean)[2]

  Gx <- list(mean = array(0, c(R, Nx, Px)))
  for (m in 1:Px) {
    Gx$mean[,,m] <- crossprod(state$Ax$mean, Kx[,,m])
  }
  Hx <- list(mean = matrix(0, R, Nx))
  for (m in 1:Px) {
    Hx$mean <- Hx$mean + state$ex$mean[m] * Gx$mean[,,m]
  }

  Gz <- list(mean = crossprod(state$Az$mean, Kz))

  F <- list(mean = crossprod(Hx$mean, Gz$mean))

  prediction <- list(Gx = Gx, Hx = Hx, Gz = Gz, F = F)
}

File: kbmf1mkl1k/kbmf1mkl1k_semisupervised_classification_variational_test.m
function prediction = kbmf1mkl1k_semisupervised_classification_variational_test(Kx, Kz, state)
    Nx = size(Kx, 2);
    Px = size(Kx, 3);
    Nz = size(Kz, 2);
    R = size(state.Ax.mu, 2);

    prediction.Gx.mu = zeros(R, Nx, Px);
    for m = 1:Px
        prediction.Gx.mu(:, :, m) = state.Ax.mu' * Kx(:, :, m);
    end
    prediction.Hx.mu = zeros(R, Nx);
    for m = 1:Px
        prediction.Hx.mu = prediction.Hx.mu + state.ex.mu(m) * prediction.Gx.mu(:, :, m);
    end

    prediction.Gz.mu = zeros(R, Nz);
    for s = 1:R
        prediction.Gz.mu(s, :) = state.Az.mu(:, s)' * Kz;
    end

    prediction.F.mu = prediction.Hx.mu' * prediction.Gz.mu;
end
